sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/UnivocityParser.scala

```diff
@@ -0,0 +1,189 @@
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.execution.datasources.csv
+
+import scala.util.control.NonFatal
+
+import com.univocity.parsers.csv.{CsvParser, CsvParserSettings}
+
+import org.apache.spark.internal.Logging
+import org.apache.spark.rdd.RDD
+import org.apache.spark.sql.catalyst.InternalRow
+import org.apache.spark.sql.catalyst.expressions.GenericMutableRow
+import org.apache.spark.sql.types.{StructField, StructType}
+
+/**
+ * Converts CSV string to a sequence of string
+ */
+private[csv] object UnivocityParser extends Logging {
+  /**
+   * Convert the input iterator to a iterator having [[InternalRow]]
+   */
+  def parseCsv(
+      iter: Iterator[String],
+      schema: StructType,
+      requiredSchema: StructType,
+      headers: Array[String],
+      shouldDropHeader: Boolean,
+      options: CSVOptions): Iterator[InternalRow] = {
+    if (shouldDropHeader) {
+      CSVUtils.dropHeaderLine(iter, options)
+    }
+    val csv = CSVUtils.filterCommentAndEmpty(iter, options)
+
+    val schemaFields = schema.fields
+    val requiredFields = requiredSchema.fields
+    val safeRequiredFields = if (options.dropMalformed) {
+      // If `dropMalformed` is enabled, then it needs to parse all the values
+      // so that we can decide which row is malformed.
+      requiredFields ++ schemaFields.filterNot(requiredFields.contains(_))
+    } else {
+      requiredFields
+    }
+    val safeRequiredIndices = new Array[Int](safeRequiredFields.length)
+    schemaFields.zipWithIndex.filter {
+      case (field, _) => safeRequiredFields.contains(field)
+    }.foreach {
+      case (field, index) => safeRequiredIndices(safeRequiredFields.indexOf(field)) = index
+    }
+    val requiredSize = requiredFields.length
+
+    tokenizeData(csv, options, headers).flatMap { tokens =>
+      if (options.dropMalformed && schemaFields.length != tokens.length) {
+        logWarning(s"Dropping malformed line: ${tokens.mkString(options.delimiter.toString)}")
+        None
+      } else if (options.failFast && schemaFields.length != tokens.length) {
+        throw new RuntimeException(s"Malformed line in FAILFAST mode: " +
+          s"${tokens.mkString(options.delimiter.toString)}")
+      } else {
+        val indexSafeTokens = if (options.permissive && schemaFields.length > tokens.length) {
+          tokens ++ new Array[String](schemaFields.length - tokens.length)
+        } else if (options.permissive && schemaFields.length < tokens.length) {
+          tokens.take(schemaFields.length)
```

Comment:
Oh, I haven't tested this yet but I am sure this will work without this logic anyway but I think it is safe to slice this here.

The size of `tokens` can be larger than `schemaFields`. I can remove this logic if you feel strongly weird but I feel like it might be okay to just leave.